pytorch_lightning/utilities/auto_restart.py

```diff
@@ -588,3 +592,96 @@ def reload_dataloader_state_dict(dataloader: DataLoader, state_dict: Dict[str, A
 
     else:
         raise MisconfigurationException("This shouldn't happen. Please, open an issue on PyTorch Lightning Github.")
+
+
+def _validate_iterable_dataset(dataloader: DataLoader) -> bool:
+
+    SUPPORTED_SAMPLERS = (RandomSampler, SequentialSampler, DistributedSampler)
+
+    dataset = dataloader.dataset
+
+    next_fn = getattr(dataset, "__next__", None)
+    if not next_fn:
+        raise MisconfigurationException(
+            "Fault Tolerant Training doesn't support an IterableDataset without a `__next__`"
+            " method implemented. Hint: We recommend you to move your logic inside and rely "
+            "on a sampler, generator to perform the iteration."
+        )
+
+    samplers = {k: v for k, v in dataset.__dict__.items() if isinstance(v, Sampler)}
+
+    if not samplers:
+        raise MisconfigurationException(
+            "Fault Tolerant Training doesn't support an IterableDataset without a sampler as attribute."
+        )
+
+    sampler = [v for v in samplers.values() if v.__class__ in SUPPORTED_SAMPLERS]
+
+    if not sampler:
+        raise MisconfigurationException(f"Fault Tolerant Training supports only {SUPPORTED_SAMPLERS}.")
+
+    if len(sampler) > 1:
+        raise MisconfigurationException(f"A single sampler is supported within an Iterable Dataset. Found {sampler}.")
+
+    if sampler[0].__class__ == DistributedSampler:
+        return not sampler.shuffle, "A `DistributedSampler` sampler shuffle attribute is set to True."
+
+    return sampler[0].__class__ == SequentialSampler, ""
+
+
+def _validate_map_dataset(dataloader: DataLoader) -> bool:
+    SUPPORTED_SAMPLERS = (RandomSampler, SequentialSampler, DistributedSampler)
+
+    sampler = getattr(dataloader, "sampler", None)
+    if sampler is not None and sampler.__class__ not in SUPPORTED_SAMPLERS:
+        raise MisconfigurationException(f"Fault Tolerant Training supports only {SUPPORTED_SAMPLERS}.")
+
+    batch_sampler = getattr(dataloader, "batch_sampler", None)
+    if batch_sampler is not None and batch_sampler.__class__ is not BatchSampler:
+        raise MisconfigurationException("Fault Tolerant Training supports only a BatchSampler.")
+
+    if sampler.__class__ == DistributedSampler:
+        return not sampler.shuffle, "A `DistributedSampler` sampler shuffle attribute is set to True."
```

Comment:
was it supposed to be an `assert`?

Reply:
No, it is the expected behaviour. I am trying to get the messaging error along to bring more identification to the users why Fault Tolerant isn't supported.

Reply:
Oh, I see. The return type annotation was misleading in that case.